[🐛 Bug]: including webdriverio as a ts dep generates a build error regarding puppeteer-core

### Have you read the Contributing Guidelines on issues?

- [X] I have read the [Contributing Guidelines on issues](https://github.com/webdriverio/webdriverio/blob/main/CONTRIBUTING.md#reporting-new-issues).

### WebdriverIO Version

8.12.1

### Node.js Version

18.15.0

### Mode

Standalone Mode

### Which capabilities are you using?

```typescript
not relevant
```


### What happened?

Trying to compile my TS based project which uses WDIO as a dependency, and getting stuck on this error:

```
node_modules/puppeteer-core/lib/esm/puppeteer/common/Debug.d.ts:22:62 - error TS7016: Could not find a declaration file for module 'debug'. '/Users/jlipps/Code/gh/jlipps/appium-ocr-plugin/node_modules/puppeteer-core/node_modules/debug/src/index.js' implicitly has an 'any' type.
  Try `npm i --save-dev @types/debug` if it exists or add a new declaration (.d.ts) file containing `declare module 'debug';`

22 export declare function importDebug(): Promise<typeof import('debug')>;
```

I'm not using `puppeteer-core` at all. It appears to be imported by WDIO.

### What is your expected behavior?

No WDIO-related TS build issues

### How to reproduce the bug.

Try to build this project at this branch: https://github.com/jlipps/appium-ocr-plugin/tree/jlipps/updates

### Relevant log output

```typescript
n/a
```


### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### Is there an existing issue for this?

- [X] I have searched the existing issues

--- a/packages/wdio-cucumber-framework/src/index.ts
+++ b/packages/wdio-cucumber-framework/src/index.ts
@@ -36,13 +36,13 @@ import {
 
     DataTable,
     World,
-    Status
+    Status,
 } from '@cucumber/cucumber'
 import Gherkin from '@cucumber/gherkin'
 import { IdGenerator } from '@cucumber/messages'
 import type { Feature, GherkinDocument } from '@cucumber/messages'
 import type Cucumber from '@cucumber/cucumber'
-import type { IRunEnvironment } from '@cucumber/cucumber/api'
+import type { IConfiguration, IRunEnvironment } from '@cucumber/cucumber/api'
 import { loadConfiguration, loadSources, runCucumber } from '@cucumber/cucumber/api'
 
 import { DEFAULT_OPTS } from './constants.js'
@@ -269,7 +269,7 @@ class CucumberAdapter {
             }
 
             const { runConfiguration } = await loadConfiguration(
-                { profiles: this._cucumberOpts.profiles, provided: this._cucumberOpts },
+                { profiles: this._cucumberOpts.profiles, provided: this._cucumberOpts as Partial<IConfiguration> },
                 environment
             )
 

--- a/packages/wdio-cucumber-framework/src/types.ts
+++ b/packages/wdio-cucumber-framework/src/types.ts
@@ -59,7 +59,7 @@ export interface CucumberOptions {
      * "random" or "random:42" whereas 42 is the seed for randomization
      * @default defined
      */
-    order?: string;
+    order?: 'defined' | 'random' | `random:${string}`;
     /**
      * Publish a report of your test run to https://reports.cucumber.io/
      */

--- a/packages/wdio-cucumber-framework/tests/adapter.test.ts
+++ b/packages/wdio-cucumber-framework/tests/adapter.test.ts
@@ -136,7 +136,7 @@ describe('CucumberAdapter', () => {
     it('can run without errors', async () => {
         const adapter = await CucumberAdapter.init!('0-0', {
             cucumberOpts: { format: [] }
-        }, ['/foo/bar'], {}, {}, {}, false, ['progress'])
+        }, ['/foo/bar'], {}, {}, {}, false, 'progress')
         adapter.registerRequiredModules = vi.fn()
         adapter.addWdioHooks = vi.fn()
         adapter.loadFiles = vi.fn()
@@ -218,7 +218,7 @@ describe('CucumberAdapter', () => {
             {},
             {},
             false,
-            ['progress']
+            'progress'
         )
         adapter.addWdioHooks(
             {
@@ -326,7 +326,7 @@ describe('CucumberAdapter', () => {
             {},
             {},
             false,
-            ['progress']
+            'progress'
         )
 
         expect(adapter._specs).toHaveLength(1)
@@ -354,7 +354,7 @@ describe('CucumberAdapter', () => {
             {},
             {},
             false,
-            ['progress']
+            'progress'
         )
 
         expect(adapter._specs).toHaveLength(1)
@@ -382,7 +382,7 @@ describe('CucumberAdapter', () => {
             {},
             {},
             false,
-            ['progress']
+            'progress'
         )
 
         expect(adapter._specs).toHaveLength(1)
@@ -410,7 +410,7 @@ describe('CucumberAdapter', () => {
             {},
             {},
             false,
-            ['progress']
+            'progress'
         )
 
         expect(adapter._specs).toHaveLength(1)
@@ -438,7 +438,7 @@ describe('CucumberAdapter', () => {
             {},
             {},
             false,
-            ['progress']
+            'progress'
         )
 
         expect(adapter._specs).toHaveLength(0)
@@ -466,7 +466,7 @@ describe('CucumberAdapter', () => {
             {},
             {},
             false,
-            ['progress']
+            'progress'
         )
 
         expect(adapter._specs).toHaveLength(1)
@@ -494,7 +494,7 @@ describe('CucumberAdapter', () => {
             {},
             {},
             false,
-            ['progress']
+            'progress'
         )
 
         expect(adapter._specs).toHaveLength(1)
@@ -521,7 +521,7 @@ describe('CucumberAdapter', () => {
             {},
             {},
             false,
-            ['progress']
+            'progress'
         )
 
         expect(adapter.gherkinParser.tokenMatcher.dialect.name).toBe('Danish')
@@ -539,7 +539,7 @@ describe('CucumberAdapter', () => {
             {},
             {},
             false,
-            ['progress']
+            'progress'
         )
 
         expect(adapter._specs).toHaveLength(2)
@@ -567,7 +567,7 @@ describe('CucumberAdapter', () => {
             {},
             {},
             false,
-            ['progress']
+            'progress'
         )
 
         expect(adapter._specs).toHaveLength(1)
@@ -594,7 +594,7 @@ describe('CucumberAdapter', () => {
             {},
             {},
             true,
-            ['progress']
+            'progress'
         )
 
         expect(adapter._specs).toHaveLength(0)

--- a/packages/wdio-utils/src/index.ts
+++ b/packages/wdio-utils/src/index.ts
@@ -6,7 +6,8 @@ import { startWebDriver } from './startWebDriver.js'
 import { initializeWorkerService, initializeLauncherService } from './initializeServices.js'
 import {
     commandCallStructure, isValidParameter, getArgumentType, safeImport,
-    isFunctionAsync, transformCommandLogResult, sleep, isAppiumCapability
+    isFunctionAsync, transformCommandLogResult, sleep, isAppiumCapability,
+    userImport
 } from './utils.js'
 import { wrapCommand, executeHooksWithArgs, executeAsync } from './shim.js'
 import * as asyncIterators from './pIteration.js'
@@ -31,6 +32,7 @@ export {
     safeImport,
     sleep,
     isAppiumCapability,
+    userImport,
     asyncIterators,
 
     /**

--- a/packages/wdio-utils/src/utils.ts
+++ b/packages/wdio-utils/src/utils.ts
@@ -122,7 +122,7 @@ export function transformCommandLogResult (result: { file?: string, script?: str
 }
 
 /**
- * checks if command argument is valid according to specificiation
+ * checks if command argument is valid according to specification
  *
  * @param  {*}       arg           command argument
  * @param  {Object}  expectedType  parameter type (e.g. `number`, `string[]` or `(number|string)`)
@@ -165,6 +165,25 @@ export function isValidParameter (arg: any, expectedType: string) {
  */
 export function getArgumentType (arg: any) {
     return arg === null ? 'null' : typeof arg
+}
+
+/**
+ * Utility to import modules with user friendly error message
+ * @param moduleName  The name of the module to import
+ * @param namedImport The name of the import to return
+ * @returns          The imported module
+ */
+export async function userImport<T> (moduleName: string, namedImport = 'default'): Promise<T> {
+    try {
+        const mod = await import(moduleName)
+        if (namedImport in mod) {
+            return mod[namedImport]
+        }
+    } catch (err) {
+        throw new Error(`Couldn't import "${moduleName}"! Do you have it installed? If not run "npm install ${moduleName}"!`)
+    }
+
+    throw new Error(`Couldn't find "${namedImport}" in module "${moduleName}"`)
 }
 
 /**

--- a/packages/wdio-utils/tests/utils.test.ts
+++ b/packages/wdio-utils/tests/utils.test.ts
@@ -1,9 +1,11 @@
+import path from 'node:path'
 import type { MockedFunction } from 'vitest'
 import { vi, describe, it, expect } from 'vitest'
 
 import {
     overwriteElementCommands, commandCallStructure, isValidParameter, definesRemoteDriver,
-    getArgumentType, isFunctionAsync, filterSpecArgs, isBase64, transformCommandLogResult
+    getArgumentType, isFunctionAsync, filterSpecArgs, isBase64, transformCommandLogResult,
+    userImport
 } from '../src/utils.js'
 
 describe('utils', () => {
@@ -228,5 +230,26 @@ describe('utils:isBase64', () => {
     it('should throw if input type not a string', () => {
         // @ts-ignore
         expect(() => isBase64(null)).toThrow('Expected string but received invalid type.')
+    })
+})
+
+describe('utils:userImport', () => {
+    it('should import module', async () => {
+        const mod = await userImport('path')
+        expect(mod).toEqual(path)
+        const join = await userImport('path', 'join')
+        expect(join).toEqual(path.join)
+    })
+
+    it('throws error message if named import not found', async () => {
+        await expect(userImport('path', 'foo'))
+            .rejects
+            .toThrow('Couldn\'t find "foo" in module "path"')
+    })
+
+    it('throws error message if module not found', async () => {
+        await expect(userImport('foobar'))
+            .rejects
+            .toThrow('Couldn\'t import "foobar"! Do you have it installed? If not run "npm install foobar"!')
     })
 })

--- a/packages/webdriverio/package.json
+++ b/packages/webdriverio/package.json
@@ -65,7 +65,8 @@
     "@types/aria-query": "^5.0.0",
     "@types/lodash.clonedeep": "^4.5.6",
     "@types/lodash.zip": "^4.2.6",
-    "got": "^12.6.0"
+    "got": "^12.6.0",
+    "puppeteer-core": "^20.9.0"
   },
   "dependencies": {
     "@types/node": "^20.1.0",
@@ -86,7 +87,6 @@
     "lodash.clonedeep": "^4.5.0",
     "lodash.zip": "^4.2.0",
     "minimatch": "^9.0.0",
-    "puppeteer-core": "^20.9.0",
     "query-selector-shadow-dom": "^1.0.0",
     "resq": "^1.9.1",
     "rgb2hex": "0.2.5",
@@ -94,10 +94,14 @@
     "webdriver": "8.29.1"
   },
   "peerDependencies": {
-    "devtools": "^8.14.0"
+    "devtools": "^8.14.0",
+    "puppeteer-core": "^20.9.0"
   },
   "peerDependenciesMeta": {
     "devtools": {
+      "optional": true
+    },
+    "puppeteer-core": {
       "optional": true
     }
   }

--- a/packages/webdriverio/src/commands/browser/getPuppeteer.ts
+++ b/packages/webdriverio/src/commands/browser/getPuppeteer.ts
@@ -1,7 +1,7 @@
-import puppeteer from 'puppeteer-core'
 import logger from '@wdio/logger'
-import type { Browser as PuppeteerBrowser } from 'puppeteer-core/lib/esm/puppeteer/api/Browser.js'
+import { safeImport } from '@wdio/utils'
 import type { Capabilities } from '@wdio/types'
+import type { Puppeteer, Browser as PuppeteerBrowser } from 'puppeteer-core'
 
 import { FF_REMOTE_DEBUG_ARG } from '../../constants.js'
 
@@ -43,6 +43,14 @@ const log = logger('webdriverio')
  * @return {PuppeteerBrowser}  initiated puppeteer instance connected to the browser
  */
 export async function getPuppeteer (this: WebdriverIO.Browser) {
+    const puppeteer = (await safeImport('puppeteer-core'))?.default as unknown as Puppeteer
+    if (!puppeteer) {
+        throw new Error(
+            'You need to install "puppeteer-core" package as a dependency ' +
+            'in order to use the "getPuppeteer" method'
+        )
+    }
+
     /**
      * check if we already connected Puppeteer and if so return
      * that instance

--- a/packages/webdriverio/src/commands/browser/getPuppeteer.ts
+++ b/packages/webdriverio/src/commands/browser/getPuppeteer.ts
@@ -1,5 +1,5 @@
 import logger from '@wdio/logger'
-import { safeImport } from '@wdio/utils'
+import { userImport } from '@wdio/utils'
 import type { Capabilities } from '@wdio/types'
 import type { Puppeteer, Browser as PuppeteerBrowser } from 'puppeteer-core'
 
@@ -43,7 +43,8 @@ const log = logger('webdriverio')
  * @return {PuppeteerBrowser}  initiated puppeteer instance connected to the browser
  */
 export async function getPuppeteer (this: WebdriverIO.Browser) {
-    const puppeteer = (await safeImport('puppeteer-core'))?.default as unknown as Puppeteer
+    const puppeteer = await userImport<Puppeteer>('puppeteer-core')
+
     if (!puppeteer) {
         throw new Error(
             'You need to install "puppeteer-core" package as a dependency ' +

--- a/packages/webdriverio/src/commands/browser/mock.ts
+++ b/packages/webdriverio/src/commands/browser/mock.ts
@@ -1,10 +1,13 @@
-import type Interception from '../../utils/interception/index.js'
+import type { CDPSession } from 'puppeteer-core'
+
 import DevtoolsNetworkInterception from '../../utils/interception/devtools.js'
 import WebDriverNetworkInterception from '../../utils/interception/webdriver.js'
+
 import { getBrowserObject } from '../../utils/index.js'
+
 import type { Mock } from '../../types.js'
 import type { MockFilterOptions } from '../../utils/interception/types.js'
-import type { CDPSession } from 'puppeteer-core/lib/esm/puppeteer/common/Connection.js'
+import type Interception from '../../utils/interception/index.js'
 
 export const SESSION_MOCKS: Record<string, Set<Interception>> = {}
 export const CDP_SESSIONS: Record<string, CDPSession> = {}

--- a/packages/webdriverio/src/types.ts
+++ b/packages/webdriverio/src/types.ts
@@ -3,7 +3,7 @@ import type { Protocol } from 'devtools-protocol'
 import type { SessionFlags, AttachOptions as WebDriverAttachOptions, BidiHandler, BidiEventHandler } from 'webdriver'
 import type { Options, Capabilities, FunctionProperties, ThenArg } from '@wdio/types'
 import type { ElementReference, ProtocolCommands } from '@wdio/protocols'
-import type { Browser as PuppeteerBrowser } from 'puppeteer-core/lib/esm/puppeteer/api/Browser.js'
+import type { Browser as PuppeteerBrowser } from 'puppeteer-core'
 
 import type * as BrowserCommands from './commands/browser.js'
 import type * as ElementCommands from './commands/element.js'

--- a/packages/webdriverio/src/utils/interception/devtools.ts
+++ b/packages/webdriverio/src/utils/interception/devtools.ts
@@ -1,7 +1,7 @@
 import fs from 'node:fs/promises'
 import path from 'node:path'
 import logger from '@wdio/logger'
-import type { CDPSession } from 'puppeteer-core/lib/esm/puppeteer/common/Connection.js'
+import type { CDPSession } from 'puppeteer-core'
 import type { Protocol } from 'devtools-protocol'
 
 import Interception from './index.js'

--- a/packages/webdriverio/src/utils/interception/types.ts
+++ b/packages/webdriverio/src/utils/interception/types.ts
@@ -1,4 +1,4 @@
-import type { CDPSession } from 'puppeteer-core/lib/esm/puppeteer/common/Connection.js'
+import type { CDPSession } from 'puppeteer-core'
 import type { JsonCompatible } from '@wdio/types'
 
 /**

--- a/packages/webdriverio/tests/commands/browser/getPuppeteer.test.ts
+++ b/packages/webdriverio/tests/commands/browser/getPuppeteer.test.ts
@@ -7,6 +7,20 @@ import { remote } from '../../../src/index.js'
 
 vi.mock('got')
 vi.mock('puppeteer-core')
+/**
+ * Given that Puppeteer is not a direct dependency of this package, we can't mock
+ * it and dynamically import it. Instead, we mock the "userImport" helper and make
+ * it resolve to the mocked Puppeteer.
+ */
+vi.mock('@wdio/utils', async (origMod) => {
+    const orig = await origMod() as any
+    // resolve the mocked puppeteer-core
+    const puppeteer = await import('puppeteer-core')
+    return {
+        ...orig,
+        userImport: vi.fn().mockResolvedValue(puppeteer.default)
+    }
+})
 vi.mock('@wdio/logger', () => import(path.join(process.cwd(), '__mocks__', '@wdio/logger')))
 
 const puppeteerConnect = vi.mocked(puppeteer.connect)

--- a/packages/webdriverio/tests/commands/browser/throttleCPU.test.ts
+++ b/packages/webdriverio/tests/commands/browser/throttleCPU.test.ts
@@ -8,6 +8,20 @@ import { remote } from '../../../src/index.js'
 
 vi.mock('got')
 vi.mock('puppeteer-core')
+/**
+ * Given that Puppeteer is not a direct dependency of this package, we can't mock
+ * it and dynamically import it. Instead, we mock the "userImport" helper and make
+ * it resolve to the mocked Puppeteer.
+ */
+vi.mock('@wdio/utils', async (origMod) => {
+    const orig = await origMod() as any
+    // resolve the mocked puppeteer-core
+    const puppeteer = await import('puppeteer-core')
+    return {
+        ...orig,
+        userImport: vi.fn().mockResolvedValue(puppeteer.default)
+    }
+})
 vi.mock('@wdio/logger', () => import(path.join(process.cwd(), '__mocks__', '@wdio/logger')))
 
 // @ts-ignore mock feature

--- a/packages/webdriverio/tests/commands/browser/throttleNetwork.test.ts
+++ b/packages/webdriverio/tests/commands/browser/throttleNetwork.test.ts
@@ -8,6 +8,20 @@ import { remote } from '../../../src/index.js'
 
 vi.mock('got')
 vi.mock('puppeteer-core')
+/**
+ * Given that Puppeteer is not a direct dependency of this package, we can't mock
+ * it and dynamically import it. Instead, we mock the "userImport" helper and make
+ * it resolve to the mocked Puppeteer.
+ */
+vi.mock('@wdio/utils', async (origMod) => {
+    const orig = await origMod() as any
+    // resolve the mocked puppeteer-core
+    const puppeteer = await import('puppeteer-core')
+    return {
+        ...orig,
+        userImport: vi.fn().mockResolvedValue(puppeteer.default)
+    }
+})
 vi.mock('@wdio/logger', () => import(path.join(process.cwd(), '__mocks__', '@wdio/logger')))
 
 // @ts-ignore mock feature

